mvc/models/model-binding.rst

```diff
@@ -1,8 +1,91 @@
+ 
+#. ``Form values``: These are form values that go in the HTTP request using the POST method. (including jQuery POST requests).
+#. ``Route values``: The set of route values provided by `routing <https://docs.asp.net/projects/mvc/en/latest/controllers/routing.html>`_. 
+#. ``Query strings``: The query string part of the URI.
+
+.. note:: Form values, route data, and query strings are all stored as name-value pairs.
+
+Since model binding asked for a key named ``id`` and there is nothing named ``id`` in the form values, it moved on to the route values looking for that key. In our example, it's a match. Binding happens, and the value is converted to the integer 2. The same request using Edit(string id) would convert to the string "2". 
+
+So far the example uses simple types. In MVC simple types are any .NET primitive type or type with a string type converter. If the action method's parameter were a class such as the ``Movie`` type, which contains both simple and class types as properties, MVC's model binding will still handle it nicely. It uses reflection and recursion to traverse the properties of class types looking for matches. Model binding looks for the pattern parameter_name.property_name to bind values to properties. If it doesn't find matching values of this form, it will attempt to bind using just the property name. For those types such as ``Collection`` types, model binding looks for matches to `parameter_name[index]` or just `[index]`. Model binding treats  ``Dictionary`` types similarly, asking for `parameter_name[key]` or just `[key]`, as long as they keys are simple types. Keys that are supported match the field names HTML and tag helpers generated for the same model type. This enables round-tripping values so that the form fields remain filled with the user's input for their convenience, for example, when bound data from a create or edit did not pass validation.
+
+In order for binding to happen, members of classes must be public, writable properties containing a default public constructor. Public fields are not bound. Properties of type ``IEnumerable`` or array must be settable and will be populated with an array. Collection properties of type ``ICollection<T>`` can be read-only. 
+
+When a parameter is bound, model binding stops looking for parameters with that name and it moves on to bind the next parameter. If binding fails, MVC does not throw an error. You can query for model state errors by checking the ``ModelState.IsValid`` method. 
+
+.. Note:: Each entry in the controller's ``ModelState`` property is a ``ModelStateEntry`` containing an ``Errors property``. It's rarely necessary to query this collection yourself. Use ``ModelState.IsValid`` instead. 
+
+Additionally, there are some special data types that MVC must consider when performing model binding:
+
+- ``IFormFile``, ``IFormFileCollection``: One or more uploaded files that are part of the HTTP request.
```

Comment:
Turns out `IFormFIleCollection` isn't supported in RC1, so let's remove it for now and file an issue to track adding it back for RC2. Instead you can bind to a collection of `IFormFile`. Proposed new text:

\`\`IFormFile\`\`, \`\`IEnumerable\<IFormFIle\>\`\`: One or more uploaded files that are part of the HTTP request.